README.md
![npm (scoped)](https://img.shields.io/npm/v/@abskmj/exchangeratesapi?label=NPM) ![NPM](https://img.shields.io/npm/l/@abskmj/exchangeratesapi?label=License) ![npm](https://img.shields.io/npm/dt/@abskmj/exchangeratesapi?label=Downloads)

# Node client for exchangeratesapi.io
Unofficial node.js client for [exchangeratesapi.io](https://exchangeratesapi.io/)

> Exchange rates API is a free service for current and historical foreign exchange rates
published by the [European Central Bank](https://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html)

# Installation

## NPM Registry
```bash
npm install --save @abskmj/exchangeratesapi
```

## Github Package
This module is also available at [Github Package](https://github.com/abskmj/exchangeratesapi/packages/222586)


# Examples
## Get all the rates on current date
```javascript
let exchange = require('@abskmj/exchangeratesapi')

const getLatestRates = async () => {
    const response = await exchange.rates()

    console.log(response.data)

    /*
      {
        rates: {
          CAD: 1.5243,
          HKD: 8.4286,
          ISK: 158.3,
          ...
        },
        base: 'EUR',
        date: '2020-05-13'
      }
    */
}
```
## Get the rates on a date
```javascript
let exchange = require('@abskmj/exchangeratesapi')

const getRatesByDate = async (date) => {
  const response = await exchange.rates({ date })

  console.log(response.data)

  /*
    {
      rates: {
        CAD: 1.4959,
        HKD: 11.2301,
        LVL: 0.7093,
        ...
      },
      base: 'EUR',
      date: '2010-01-12'
    }
  */
}
```

## Get the rates of a base currency
```javascript
let exchange = require('@abskmj/exchangeratesapi')

const getLatestRatesForBaseCurrency = async(base) => {
  const response = await exchange.rates({ base })

  console.log(response.data)

  /*
    {
      rates: {
        CAD: 1.4016551724,
        HKD: 7.7504367816,
        ISK: 145.5632183908,
        ...
      },
      base: 'USD',
      date: '2020-05-13'
    }
  */
}
```

## Get specific rates of a base currency
```javascript
let exchange = require('@abskmj/exchangeratesapi')

const getLatestRatesForCurrencies = async (symbols) => {
  const response = await exchange.rates({ symbols })

  console.log(response.data)

  /*
    {
      rates: { USD: 1.0875, GBP: 0.88245 },
      base: 'EUR',
      date: '2020-05-13'
    }
  */
}
```

## Get historical rates
```javascript
let exchange = require('@abskmj/exchangeratesapi')

const getHistoricalRates = async(start_at, end_at) => {
  const response = await exchange.history({ start_at, end_at})

  console.log(response.data)
}
```

# Works with Promises and Callbacks
```javascript
// promise
exchange().then(response =>{
    console.log(response.data);
});

// callback
exchange({}, (err, response) => {
    if(err) return console.error(err)

    console.log(response.data)
})
```

# API Documentation
Documentation is available at [API.md](API.md)

# Changelog
Change log is available at [CHANGELOG.md](CHANGELOG.md)

CHANGELOG.md
# Changelog

## [2.0.0]
### Removed
- Remove `.getClient()` method
### Added
- Add `.rates()` method for exchange rates
- Add `.history()` method for historical rates
- Add Standard.js for linting
- Add Github action to publish to NPM Registry and Github Packages

## [1.1.0]
### Fixed
- Fix API base url from `https://exchangeratesapi.io/api` to `https://api.exchangeratesapi.io`
- Fix NPM audit issues

## [1.0.0]
### Added
- Add methods to call apis at [exchangeratesapi.io](https://exchangeratesapi.io/)
- Add test cases

test.js
const { expect } = require('chai')
const exchange = require('./index')

describe('Client', function () {
  it('should return all rates of current date', function () {
    return exchange.rates().then(response => {
      expect(response.status).to.equals(200)
      expect(response.request.url).to.equals('https://api.exchangeratesapi.io/latest')
    })
  })

  it('should return all rates of a particular date', function () {
    const particular = '2018-01-01'
    return exchange.rates({ date: particular }).then(response => {
      expect(response.status).to.equals(200)
      expect(response.request.url).to.equals(`https://api.exchangeratesapi.io/${particular}`)
    })
  })

  it('should return all rates of a particular currency', function () {
    const particular = 'USD'
    return exchange.rates({ base: particular }).then(response => {
      expect(response.status).to.equals(200)
      expect(response.data.base).to.equals(particular)
    })
  })

  it('should return particular rates of a particular currency', function () {
    const particular = 'USD'
    return exchange.rates({
      base: particular,
      symbols: 'EUR,GBP'
    }).then(response => {
      expect(response.status).to.equals(200)
      expect(response.data.base).to.equals(particular)
      expect(response.data.rates).to.have.property('EUR')
      expect(response.data.rates).to.not.have.property('AUD')
    })
  })

  it('should return historical rates for a time period', function () {
    const startDate = '2018-01-01'
    const endDate = '2018-01-15'

    return exchange.history({
      start_at: startDate,
      end_at: endDate
    }).then(response => {
      expect(response.status).to.equals(200)
      expect(response.request.url).to.equals(`https://api.exchangeratesapi.io/history?end_at=${endDate}&start_at=${startDate}`)
      expect(response.data.start_at).to.equals(startDate)
      expect(response.data.end_at).to.have.equals(endDate)
    })
  })

  it('should return historical rates for specific currencies', async () => {
    const startDate = '2018-01-01'
    const endDate = '2018-01-15'

    const response = await exchange.history({
      start_at: startDate,
      end_at: endDate,
      symbols: 'ILS,JPY'
    })

    expect(response.status).to.equals(200)
    expect(response.request.url).to.equals(`https://api.exchangeratesapi.io/history?end_at=${endDate}&start_at=${startDate}&symbols=ILS%2CJPY`)
    expect(response.data.start_at).to.equals(startDate)
    expect(response.data.end_at).to.have.equals(endDate)
  })

  it('should return historical rate for specific base currency', async () => {
    const startDate = '2018-01-01'
    const endDate = '2018-01-15'
    const base = 'USD'

    const response = await exchange.history({
      start_at: startDate,
      end_at: endDate,
      base
    })

    expect(response.status).to.equals(200)
    expect(response.request.url).to.equals(`https://api.exchangeratesapi.io/history?base=${base}&end_at=${endDate}&start_at=${startDate}`)
    expect(response.data.start_at).to.equals(startDate)
    expect(response.data.end_at).to.have.equals(endDate)
    expect(response.data.base).to.have.equals(base)
  })
})


index.js
/**
 * Unofficial node.js client for exchangeratesapi.io
 *
 * Import the client
 * ```
 * const exchange = require('@abskmj/exchangeratesapi')
 * ```
 * @namespace exchange
 */
module.exports = require('@abskmj/cligen').getClient({
  baseUrl: 'https://api.exchangeratesapi.io',
  operations: {
    /**
     * Get exchange rates
     * @function rates
     * @memberof exchange
     * @param {string} [date=latest] Date in YYYY-MM-DD format
     * @param {string} [base=EUR] Currency against which rates are quoted
     * @param {string} [symbols] Comma separated currencies for which rates are needed
     * @returns {Axios.Response} [Response](https://github.com/axios/axios#response-schema) from Axios module
     * @example
     * exchange.rates()
     * @example
     * exchange.rates({ base: 'USD' })
     * @example
     * exchange.rates({ symbols: 'USD,GBP' })
     * @example
     * exchange.rates({ date: '2010-01-12' })
     */
    rates: {
      uri: '/{date}',
      data: {
        path: {
          parameters: [{
            name: 'date',
            default: 'latest'
          }]
        },
        query: {
          parameters: [
            { name: 'base' },
            { name: 'symbols' }
          ]
        }
      }
    },
    /**
     * Get historical exchange rates
     * @function history
     * @memberof exchange
     * @param {string} start_at State date in YYYY-MM-DD format
     * @param {string} end_at End date in YYYY-MM-DD format
     * @param {string} [base=EUR] Currency against which rates are quoted
     * @param {string} [symbols] Comma separated currencies for which rates are needed
     * @returns {Axios.Response} [Response](https://github.com/axios/axios#response-schema) from Axios module
     * @example
     * exchange.history()
     * @example
     * exchange.history({ base: 'USD' })
     * @example
     * exchange.history({ symbols: 'USD,GBP' })
     * @example
     * exchange.history({ date: '2010-01-12' })
     */
    history: {
      uri: '/history',
      data: {
        query: {
          parameters: [
            { name: 'start_at' },
            { name: 'end_at' },
            { name: 'base' },
            { name: 'symbols' }
          ]
        }
      }
    }
  }
})


.examples/api.js
const exchange = require('../index')

const getLatestRates = async () => {
    const response = await exchange.rates()

    console.log(response.data)

    /*
      {
        rates: {
          CAD: 1.5243,
          HKD: 8.4286,
          ISK: 158.3,
          PHP: 54.545,
          DKK: 7.4556,
          HUF: 353.58,
          CZK: 27.408,
          AUD: 1.6687,
          RON: 4.8353,
          SEK: 10.5843,
          IDR: 16092.26,
          INR: 81.8825,
          BRL: 6.3606,
          RUB: 79.5893,
          HRK: 7.5705,
          JPY: 116.28,
          THB: 34.844,
          CHF: 1.0528,
          SGD: 1.538,
          PLN: 4.5636,
          BGN: 1.9558,
          TRY: 7.5861,
          CNY: 7.7102,
          NOK: 10.938,
          NZD: 1.7983,
          ZAR: 19.919,
          USD: 1.0875,
          MXN: 26.2304,
          ILS: 3.819,
          GBP: 0.88245,
          KRW: 1331.08,
          MYR: 4.704
        },
        base: 'EUR',
        date: '2020-05-13'
      }
    */
}

const getLatestRatesForBaseCurrency = async(base) => {
  const response = await exchange.rates({ base })

  console.log(response.data)

  /*
    {
      rates: {
        CAD: 1.4016551724,
        HKD: 7.7504367816,
        ISK: 145.5632183908,
        PHP: 50.1563218391,
        DKK: 6.8557241379,
        HUF: 325.1310344828,
        CZK: 25.2027586207,
        GBP: 0.8114482759,
        RON: 4.4462528736,
        SEK: 9.7326896552,
        IDR: 14797.4804597701,
        INR: 75.2942528736,
        BRL: 5.8488275862,
        RUB: 73.1855632184,
        HRK: 6.9613793103,
        JPY: 106.924137931,
        THB: 32.0404597701,
        CHF: 0.968091954,
        EUR: 0.9195402299,
        MYR: 4.3255172414,
        BGN: 1.7984367816,
        TRY: 6.9757241379,
        CNY: 7.0898390805,
        NOK: 10.0579310345,
        NZD: 1.6536091954,
        ZAR: 18.3163218391,
        USD: 1,
        MXN: 24.119908046,
        SGD: 1.4142528736,
        AUD: 1.5344367816,
        ILS: 3.5117241379,
        KRW: 1223.9816091954,
        PLN: 4.1964137931
      },
      base: 'USD',
      date: '2020-05-13'
    }
  */
}

const getLatestRatesForCurrencies = async (symbols) => {
  const response = await exchange.rates({ symbols })

  console.log(response.data)

  /*
    {
      rates: { USD: 1.0875, GBP: 0.88245 },
      base: 'EUR',
      date: '2020-05-13'
    }
  */
}

const getRatesByDate = async (date) => {
  const response = await exchange.rates({ date })

  console.log(response.data)

  /*
    {
      rates: {
        CAD: 1.4959,
        HKD: 11.2301,
        LVL: 0.7093,
        PHP: 66.106,
        DKK: 7.4405,
        HUF: 268.18,
        CZK: 26.258,
        AUD: 1.5668,
        RON: 4.1405,
        SEK: 10.2215,
        IDR: 13281.14,
        INR: 66.21,
        BRL: 2.5309,
        RUB: 42.6974,
        LTL: 3.4528,
        JPY: 132.41,
        THB: 47.839,
        CHF: 1.4743,
        SGD: 2.0133,
        PLN: 4.0838,
        BGN: 1.9558,
        TRY: 2.1084,
        CNY: 9.8863,
        NOK: 8.1825,
        NZD: 1.9573,
        ZAR: 10.8264,
        USD: 1.4481,
        MXN: 18.4995,
        EEK: 15.6466,
        GBP: 0.8972,
        KRW: 1627.4,
        MYR: 4.8424,
        HRK: 7.2753
      },
      base: 'EUR',
      date: '2010-01-12'
    }
  */
}

const getHistoricalRates = async(start_at, end_at) => {
  const response = await exchange.history({ start_at, end_at})

  console.log(response.data)
}

getLatestRates()
getLatestRatesForBaseCurrency('USD')
getLatestRatesForCurrencies('USD,GBP')
getRatesByDate('2010-01-12')
getHistoricalRates('2018-01-01', '2018-01-31')